ethdb/olddb/memorymutation.go

```diff
@@ -183,109 +183,155 @@ func (m *miningmutation) Has(table string, key []byte) (bool, error) {
+func (m *memorymutation) Close() {
 	m.Rollback()
 }
 
-func (m *miningmutation) Begin(ctx context.Context, flags ethdb.TxFlags) (ethdb.DbWithPendingMutations, error) {
+func (m *memorymutation) Begin(ctx context.Context, flags ethdb.TxFlags) (ethdb.DbWithPendingMutations, error) {
 	panic("mutation can't start transaction, because doesn't own it")
 }
 
-func (m *miningmutation) panicOnEmptyDB() {
+func (m *memorymutation) panicOnEmptyDB() {
 	if m.db == nil {
 		panic("Not implemented")
 	}
 }
 
-func (m *miningmutation) SetRwKV(kv kv.RwDB) {
+func (m *memorymutation) SetRwKV(kv kv.RwDB) {
 	m.db.(ethdb.HasRwKV).SetRwKV(kv)
 }
 
-func (m *miningmutation) BucketSize(bucket string) (uint64, error) {
+func (m *memorymutation) BucketSize(bucket string) (uint64, error) {
 	return 0, nil
 }
 
-func (m *miningmutation) DropBucket(bucket string) error {
+func (m *memorymutation) DropBucket(bucket string) error {
 	panic("Not implemented")
 }
 
-func (m *miningmutation) ExistsBucket(bucket string) (bool, error) {
+func (m *memorymutation) ExistsBucket(bucket string) (bool, error) {
 	panic("Not implemented")
 }
 
-func (m *miningmutation) ListBuckets() ([]string, error) {
+func (m *memorymutation) ListBuckets() ([]string, error) {
 	panic("Not implemented")
 }
 
-func (m *miningmutation) ClearBucket(bucket string) error {
+func (m *memorymutation) ClearBucket(bucket string) error {
 	m.clearedTables[bucket] = struct{}{}
 	return m.memTx.ClearBucket(bucket)
 }
 
-func (m *miningmutation) isBucketCleared(bucket string) bool {
+func (m *memorymutation) isBucketCleared(bucket string) bool {
 	_, ok := m.clearedTables[bucket]
 	return ok
 }
 
-func (m *miningmutation) CollectMetrics() {
+func (m *memorymutation) CollectMetrics() {
 }
 
-func (m *miningmutation) CreateBucket(bucket string) error {
+func (m *memorymutation) CreateBucket(bucket string) error {
 	panic("Not implemented")
 }
 
+func (m *memorymutation) Flush(tx kv.RwTx) error {
+	// Obtain buckets touched.
+	buckets, err := m.memTx.ListBuckets()
+	if err != nil {
+		return err
+	}
+	// Iterate over each bucket and apply changes accordingly.
+	for _, bucket := range buckets {
+		if _, ok := m.dupsortTables[bucket]; ok && bucket != kv.HashedStorage {
+			cbucket, err := m.memTx.CursorDupSort(bucket)
+			if err != nil {
+				return err
+			}
+			defer cbucket.Close()
+			dbCursor, err := tx.RwCursorDupSort(bucket)
+			if err != nil {
+				return err
+			}
+			defer dbCursor.Close()
+			for k, v, err := cbucket.First(); k != nil; k, v, err = cbucket.Next() {
+				if err != nil {
+					return err
+				}
+				if err := dbCursor.AppendDup(k, v); err != nil {
+					return err
+				}
+			}
+		} else {
+			cbucket, err := m.memTx.Cursor(bucket)
+			if err != nil {
+				return err
+			}
+			defer cbucket.Close()
+			for k, v, err := cbucket.First(); k != nil; k, v, err = cbucket.Next() {
+				if err != nil {
+					return err
+				}
+				if err := tx.Put(bucket, k, v); err != nil {
```

Comment:
FYI: 
- in the past we thought that "Append" producing more compact tables inside DB than "Put". Actually not, it was bug in MDBX. Now Append and "Sorted puts" producing tables of same "compactness".
- Append can be a little faster than "sorted puts", but profit is small - better go for simplicity here 
- so, just avoid "unsorted puts" is enough

Reply:
i think